plugins/bil_pipe/bil_const_propagation.ml

```diff
@@ -0,0 +1,236 @@
+        | _ -> Some Top)
+
+  let defs bil =
+    (object
+      inherit [Var.Set.t] Stmt.visitor
+      method! enter_move v _ defs = Set.add defs v
+    end)#run bil Var.Set.empty |> Set.to_list
+
+  let has_jmps bil =
+    List.exists bil
+      ~f:(function | Jmp _ -> true | _ -> false)
+
+  class t = object(self)
+    inherit e
+
+    method eval_stmt (s : stmt) : stmt r =
+      match s with
+      | Move (v,e)  -> self#eval_move v e
+      | If (c,y,n)  -> self#eval_if c y n
+      | While (c,b) -> self#eval_while c b
+      | Jmp e -> self#eval_jmp e
+      | _ -> SM.return s
+
+    method eval (bil : bil) : bil r =
+      SM.List.fold bil ~init:[] ~f:(fun acc s ->
+          self#eval_stmt s >>= fun s ->
+          SM.return (s :: acc)) >>= fun bil ->
+      SM.return (List.rev bil)
+
+    method eval_move v e : stmt r =
+      match e with
+      | Int w ->
+        self#update (Def w) v >>= fun () ->
+        SM.return (Bil.move v e)
+      | e ->
+        self#eval_exp e >>= fun e ->
+        self#update Top v >>= fun () ->
+        SM.return (Bil.move v e)
+
+    method eval_if cond yes no =
+      SM.get () >>= fun ctxt ->
+      self#eval_exp cond >>= fun cond ->
+      self#eval yes >>= fun yes ->
+      SM.get () >>= fun ctxt_yes ->
+      SM.put ctxt >>= fun () ->
+      self#eval no >>= fun no ->
+      SM.get () >>= fun ctxt_no ->
+      let ctxt = match has_jmps yes, has_jmps no with
+        | false, false -> meet ctxt_yes ctxt_no
+        | true, true -> ctxt
+        | false, _ -> ctxt_yes
+        | _ -> ctxt_no in
+      SM.put ctxt >>= fun () ->
+      SM.return (Bil.if_ cond yes no)
+
+    method eval_while cond body =
+      SM.get () >>= fun ctxt ->
+      SM.List.iter (defs body) ~f:(self#update Top) >>= fun () ->
+      self#eval_exp cond >>= fun cond ->
+      self#eval body >>= fun body ->
+      SM.get () >>= fun ctxt' ->
+      let ctxt = if has_jmps body then ctxt
+        else meet ctxt ctxt' in
+      SM.put ctxt >>= fun () ->
+      SM.return (Bil.while_ cond body)
+
+    method eval_jmp e : stmt r =
+      self#eval_exp e >>= fun e ->
+      SM.return (Bil.jmp e)
+  end
+end
+
+module Dead_code = struct
+
+  let is_used var bil =
+    let (||) x y = match x, y with
+      | Some r1, Some r2 -> Some (r1 || r2)
+      | Some _, None | None, Some _ -> Some true
+      | None, None -> None in
+    let is_in_free s = Set.mem (Stmt.free_vars s) var in
+    let defined = Var.equal var in
+    let rec used_in = function
+      | [] -> None
+      | s :: bil -> match s with
+        | If (_, yes, no) ->
+          let r = used_in yes || used_in no in
+          if Option.is_some r then r
+          else used_in bil
+        | While (_, body) -> used_in body || used_in bil
+        | Move (var',_) when defined var' -> Some false
+        | s when is_in_free s -> Some true
+        | _ -> used_in bil in
+    match used_in bil with
+    | None -> false
+    | Some x -> x
+
+  let eliminate bil =
```

Comment:
This is an extremely suboptimal implementation, for each statement you look all whether its variable is used in all consecutive statements, which is O(n^2). You can implement this in O(n), if you will first translate each stmt list, into `(stmt * live) list`.